src/finiteVolume/processes/twoPhaseEulerAdvection.cpp

```diff
@@ -203,29 +203,99 @@ ablate::finiteVolume::processes::TwoPhaseEulerAdvection::TwoPhaseEulerAdvection(
     }
     if (flow.GetSubDomain().ContainsField(CompressibleFlowFields::TEMPERATURE_FIELD)) {
         // add in aux update variables
-        flow.RegisterAuxFieldUpdate(UpdateAuxTemperatureField2Gas, this, std::vector<std::string>{CompressibleFlowFields::TEMPERATURE_FIELD}, {"densityVF", "euler"});
+        flow.RegisterAuxFieldUpdate(UpdateAuxTemperatureField2Gas, this, std::vector<std::string>{CompressibleFlowFields::TEMPERATURE_FIELD}, {"volumeFraction", "densityVF", "euler"});
     }
     if (flow.GetSubDomain().ContainsField("pressure")) {
         // add in aux update variables
-        flow.RegisterAuxFieldUpdate(UpdateAuxPressureField2Gas, this, std::vector<std::string>{"pressure"}, {"densityVF", "euler"});
+        flow.RegisterAuxFieldUpdate(UpdateAuxPressureField2Gas, this, std::vector<std::string>{"pressure"}, {"volumeFraction", "densityVF", "euler"});
     }
-    if (flow.GetSubDomain().ContainsField("volumeFraction")) {
-        flow.RegisterAuxFieldUpdate(UpdateAuxVolumeFractionField2Gas, this, std::vector<std::string>{"volumeFraction"}, {"densityVF", "euler"});
+}
+PetscErrorCode ablate::finiteVolume::processes::TwoPhaseEulerAdvection::MultiphaseFlowPreStage(TS flowTs, ablate::solver::Solver &solver, PetscReal stagetime) {
+    PetscFunctionBegin;
+    // Get flow field data
+    const auto &fvSolver = dynamic_cast<ablate::finiteVolume::FiniteVolumeSolver &>(solver);
+    solver::Range cellRange;
+    fvSolver.GetCellRangeWithoutGhost(cellRange);
+    PetscInt dim;
+    PetscCall(DMGetDimension(fvSolver.GetSubDomain().GetDM(), &dim));
+    const auto &flowEulerId = fvSolver.GetSubDomain().GetField("euler").id;  // constant instead of string for euler
+    auto &vfEulerId = fvSolver.GetSubDomain().GetField("volumeFraction").id;
+    DM dm = fvSolver.GetSubDomain().GetDM();
+    Vec globFlowVec;
+    PetscCall(TSGetSolution(flowTs, &globFlowVec));
+
+    const PetscScalar *flowArray;
+    PetscScalar *flowArrayV2;
+    PetscCall(VecGetArrayRead(globFlowVec, &flowArray));
+    PetscCall(VecGetArray(globFlowVec, &flowArrayV2));
+
+    for (PetscInt i = cellRange.start; i < cellRange.end; ++i) {
+        const PetscInt cell = cellRange.points ? cellRange.points[i] : i;
+        const PetscScalar *eulerField = nullptr;                                                 // not const if modifying value
+        DMPlexPointLocalFieldRead(dm, cell, flowEulerId, flowArray, &eulerField) >> checkError;  // to write -read or ref
+        auto density = eulerField[ablate::finiteVolume::CompressibleFlowFields::RHO];
+        PetscReal velocity[3];
+        for (PetscInt d = 0; d < dim; d++) {
+            velocity[d] = eulerField[ablate::finiteVolume::CompressibleFlowFields::RHOU + d] / density;
+        }
+
+        PetscScalar *vfField = nullptr;
+        DMPlexPointLocalFieldRef(dm, cell, vfEulerId, flowArrayV2, &vfField) >> checkError;
+
+        // reproducing uOff
+        PetscInt uOff[3];
+        uOff[0] = 3 + dim;  // alpha
```

Comment:
I don't know if this is always true.  I think we need to get this from the fields.